go/apps/mixnet/mixnet.go

```diff
@@ -0,0 +1,139 @@
+// Copyright (c) 2015, Google Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mixnet
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"net"
+
+	"github.com/jlmucb/cloudproxy/go/tao"
+)
+
+const (
+	CellBytes = 1024 // Length of a cell
+)
+
+// x509 identity of a mixnet router.
+var x509Identity pkix.Name = pkix.Name{
+	Organization:       []string{"Google Inc."},
+	OrganizationalUnit: []string{"Cloud Security"},
+}
+
+// RouterContext stores the runtime environment for a Tao-delegated router.
+type RouterContext struct {
+	Keys     *tao.Keys    // Signing keys of this hosted program.
+	Domain   *tao.Domain  // Policy guard and public key.
+	Listener net.Listener // Socket where server listens for clients.
+}
+
+// NewRouterContext generates new keys, loads a local domain configuration from
+// `path`, and binds an anonymous listener socket to `addr` on network
+// `network`. It is expected that this program will be called from a host
+// implementing the Tao, as it requests a delegation from its parent.
+func NewRouterContext(path, network, addr string) (hp *RouterContext, err error) {
+	hp = new(RouterContext)
+
+	// Generate keys and get attestation from parent.
+	hp.Keys, err = tao.NewTemporaryTaoDelegatedKeys(tao.Signing|tao.Crypting, tao.Parent())
+	if err != nil {
+		return nil, err
+	}
+
+	// Create a certificate.
+	hp.Keys.Cert, err = hp.Keys.SigningKey.CreateSelfSignedX509(&x509Identity)
```

Comment:
This identity should be configurable, since callers of this library might not want to be Google/Cloud Security. So, this should be passed as an argument to NewRouterContext.